realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp

```diff
@@ -0,0 +1,209 @@
+ jboolean cache, jboolean disable_format_upgrade, jboolean auto_change_notification) {
+    TR_ENTER()
+    JStringAccessor path(env, realm_path);
+    Realm::Config *config = new Realm::Config();
+    config->path = path;
+    if (key != nullptr) {
+        config->encryption_key = jbytearray_to_vector(env, key);
+    }
+    config->read_only = read_only;
+    config->in_memory = in_memory;
+    config->cache = cache;
+    config->disable_format_upgrade = disable_format_upgrade;
+    config->automatic_change_notifications = auto_change_notification;
+    return reinterpret_cast<jlong>(config);
+}
+
+JNIEXPORT void JNICALL
+Java_io_realm_internal_SharedRealm_nativeCloseConfig
+(JNIEnv *, jclass, jlong config_ptr) {
+    TR_ENTER_PTR(config_ptr)
+    delete RC(config_ptr);
+}
+
+JNIEXPORT jlong JNICALL
+Java_io_realm_internal_SharedRealm_nativeGetSharedRealm
+(JNIEnv *, jclass, jlong config_ptr) {
+    TR_ENTER_PTR(config_ptr)
+    SharedRealm shared_realm = Realm::get_shared_realm(*RC(config_ptr));
+    return reinterpret_cast<jlong>(new SharedRealm(shared_realm));
+}
+
+JNIEXPORT void JNICALL
+Java_io_realm_internal_SharedRealm_nativeCloseSharedRealm
+(JNIEnv *, jclass, jlong shared_realm_ptr) {
+    TR_ENTER_PTR(shared_realm_ptr)
+    delete SR(shared_realm_ptr);
+}
+
+JNIEXPORT void JNICALL
+Java_io_realm_internal_SharedRealm_nativeBeginTransaction
+(JNIEnv *, jclass, jlong shared_realm_ptr) {
+    TR_ENTER_PTR(shared_realm_ptr)
+    (*SR(shared_realm_ptr))->begin_transaction();
+}
+
+JNIEXPORT void JNICALL
+Java_io_realm_internal_SharedRealm_nativeCommitTransaction
+(JNIEnv *, jclass, jlong shared_realm_ptr) {
+    TR_ENTER_PTR(shared_realm_ptr)
+    (*SR(shared_realm_ptr))->commit_transaction();
+}
+
+JNIEXPORT void JNICALL
+Java_io_realm_internal_SharedRealm_nativeCancelTransaction
+(JNIEnv *, jclass, jlong shared_realm_ptr) {
+    TR_ENTER_PTR(shared_realm_ptr)
+    (*SR(shared_realm_ptr))->cancel_transaction();
+}
+
+
+JNIEXPORT jboolean JNICALL
+Java_io_realm_internal_SharedRealm_nativeIsInTransaction(JNIEnv *, jclass, jlong shared_realm_ptr) {
+    TR_ENTER_PTR(shared_realm_ptr)
+    return (*SR(shared_realm_ptr))->is_in_transaction();
+}
+
+JNIEXPORT jlong JNICALL
+Java_io_realm_internal_SharedRealm_nativeReadGroup
+(JNIEnv *, jclass , jlong shared_realm_ptr) {
+    TR_ENTER_PTR(shared_realm_ptr)
+    return reinterpret_cast<jlong>((*SR(shared_realm_ptr))->read_group());
+}
+
+JNIEXPORT jlong JNICALL
+Java_io_realm_internal_SharedRealm_nativeGetVersion
+(JNIEnv *, jclass, jlong shared_realm_ptr) {
+    TR_ENTER_PTR(shared_realm_ptr)
+    return ObjectStore::get_schema_version((*SR(shared_realm_ptr))->read_group());
+}
+
+JNIEXPORT jboolean JNICALL
+Java_io_realm_internal_SharedRealm_nativeIsEmpty(JNIEnv *, jclass, jlong shared_realm_ptr) {
+    TR_ENTER_PTR(shared_realm_ptr)
+    return JB(ObjectStore::is_empty((*SR(shared_realm_ptr))->read_group()));
```

Comment:
I think we should have a inlined function called `to_jboolean` instead of `JB`. We already have `to_jstring()` and the upper case macros are used for casting of pointers.